core/src/sumeragi/view_change.rs

```diff
@@ -16,90 +16,50 @@ use crate::block::{EmptyChainHash, VersionedCommittedBlock, VersionedValidBlock}
 
-    /// Constructor for `BlockCreationTimeout` view change suggestion.
-    /// # Errors
-    /// Can fail due to signing
-    pub fn block_creation_timeout(
-        previous_proof: HashOf<Self>,
-        latest_block: HashOf<VersionedCommittedBlock>,
-        key_pair: KeyPair,
-    ) -> Result<Self> {
-        let payload = ProofPayload {
-            reason: Reason::from(BlockCreationTimeout),
-            previous_proof,
-            latest_block,
-        };
-        Self::from_payload(payload, key_pair)
-    }
-
-    /// Constructor for `NoTransactionReceiptReceived` view change suggestion.
-    /// # Errors
-    /// Can fail due to signing
-    pub fn no_transaction_receipt_received(
-        previous_proof: HashOf<Self>,
-        latest_block: HashOf<VersionedCommittedBlock>,
-        key_pair: KeyPair,
-    ) -> Result<Self> {
-        let payload = ProofPayload {
-            reason: Reason::NoTransactionReceiptReceived(NoTransactionReceiptReceived),
-            previous_proof,
-            latest_block,
-        };
-        Self::from_payload(payload, key_pair)
+        let mut buf = [0_u8; Hash::LENGTH + std::mem::size_of::<u64>()];
+        let prev_block_hash = self.latest_block_hash.as_ref();
+        for i in 0..Hash::LENGTH {
+            buf[i] = prev_block_hash[i];
+        }
+        {
+            let mut final_buf_segment = &mut buf[Hash::LENGTH..];
+            LittleEndian::write_u64(final_buf_segment, self.view_change_index);
+        }
+        // Now we hash the buffer to produce a payload that is completely
+        // different between view change proofs in the same sumeragi round.
+        Hash::new(buf)
     }
 
     /// Signs this message with the peer's public and private key.
     /// This way peers vote for changing the view - changing the roles of peers.
     ///
     /// # Errors
     /// Can fail during creation of signature
-    pub fn sign(mut self, key_pair: KeyPair) -> Result<Self> {
-        let signature = SignatureOf::new(key_pair, &self.payload)?.transmute();
-        self.signatures.insert(signature);
-        Ok(self)
+    pub fn sign(&mut self, key_pair: KeyPair) -> Result<()> {
+        let signature = Signature::new(key_pair, self.signature_payload().as_ref())?;
+        self.signatures.push(signature);
+        Ok(())
     }
 
-    /// Adds verified signatures of `other` to self.
-    pub fn merge_signatures(&mut self, other: SignaturesOf<Proof>) {
-        self.signatures
-            .extend(other.into_verified_by_hash(&self.hash()))
+    /// Verifies and adds signatures of other to this proof.
+    pub fn merge_signatures(&mut self, other: &Vec<Signature>) {
+        let signature_payload = self.signature_payload();
+        for signature in other {
+            if signature.verify(signature_payload.as_ref()).is_ok()
+                && !self.signatures.contains(signature)
+            {
+                self.signatures.push(signature.clone());
+            }
+        }
```

Comment:
To me it seems that this is a library function.

Reply:
Is it really that bad to have an array of something?